sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts

```diff
@@ -673,17 +707,208 @@ export class EducationProgramOfferingService extends RecordDataModelService<Educ
+      applications = await this.getApplicationsToSubmitReassessment(offeringId);
+
+      for (const application of applications) {
+        application.currentAssessment = {
+          application: { id: application.id } as Application,
+          triggerType: AssessmentTriggerType.OfferingChange,
+          offering: { id: offeringId } as EducationProgramOffering,
+          creator: auditUser,
+          createdAt: currentDate,
+          submittedBy: auditUser,
+          submittedDate: currentDate,
+        } as StudentAssessment;
+        application.modifier = auditUser;
+        application.updatedAt = currentDate;
+      }
+    } else {
+      requestedOffering.offeringStatus = OfferingStatus.ChangeDeclined;
+      precedingOffering.offeringStatus = OfferingStatus.Approved;
+    }
+
+    await this.dataSource.transaction(async (transactionalEntityManager) => {
+      // create the note for assessment.
+      const note = new Note();
+      note.description = assessmentNotes;
+      note.noteType = NoteType.Program;
+      note.creator = auditUser;
+      const noteEntity = await transactionalEntityManager
+        .getRepository(Note)
+        .save(note);
+
+      // update note.
+      requestedOffering.offeringNote = noteEntity;
+
+      // update institution note.
+      await transactionalEntityManager
+        .getRepository(Institution)
+        .createQueryBuilder()
+        .relation(Institution, "notes")
+        .of({
+          id: requestedOffering.institutionLocation.institution.id,
+        } as Institution)
+        .add(noteEntity);
+
+      // Save the requested and preceding offering.
+      await transactionalEntityManager
+        .getRepository(EducationProgramOffering)
+        .save([requestedOffering, precedingOffering]);
+
+      // Save applications with new current assessment on approval.
+      if (applications && applications.length > 0) {
+        await transactionalEntityManager
+          .getRepository(Application)
+          .save(applications);
+      }
+    });
+
+    // Once the impacted applications are updated with new current assessment
+    // start the assessment workflows and delete the existing workflow instances.
+    if (applications && applications.length > 0) {
+      await this.startOfferingChangeAssessments(applications);
+    }
+  }
+
+  /**
+   * For an offering change start the assessment workflows for all new assessments
+   * and delete the existing workflow instances of previous assessments.
+   * @param applications applications impacted by offering change.
+   */
+  private async startOfferingChangeAssessments(
+    applications: ApplicationAssessmentSummary[],
+  ): Promise<void> {
+    const promises: Promise<void | WorkflowStartResult>[] = [];
+    // Used to limit the number of asynchronous operations
+    // that will start at the same time based on length of cpus.
+    const maxPromisesAllowed = os.cpus().length;
```

Comment:
We talked about having some global variable (e.g. MAX_DEGREE_OF_PARALLELISM) to avoid flooding the Camunda server with requests. Any particular reason why we are using the CPU cores here?